### ISSUE TITLE
VirtualizingStackPanel render only first item

### ISSUE THREAD

**Issue by Meloman19** [AvaloniaUI/Avalonia#12432]
**Describe the bug**
VirtualizingStackPanel render only first item when move from out of bounds with rendertransform.

**To Reproduce**
1. Create blank project.
<details>
  <summary>2. On MainView.axaml insert:</summary>
  
```xml
<Border Width="300"
		ClipToBounds="True">
	<Grid RowDefinitions="Auto, *">
		<UniformGrid Rows="1">
			<Button Content="Left"
					Click="Left_Click"
					HorizontalAlignment="Center"/>
			<Button Content="Right"
					Click="Right_Click"
					HorizontalAlignment="Center"/>
		</UniformGrid>
			<StackPanel x:Name="Panel"
					Orientation="Horizontal"
					Grid.Row="1"
					Spacing="10">
			<Border Background="Beige"
					Width="300">
				<ItemsControl Theme="{StaticResource ItemsControlTheme}">
					<TextBlock Text="1111" />
					<TextBlock Text="2222" />
					<TextBlock Text="3333" />
				</ItemsControl>
			</Border>
			<Border Background="LightBlue"
					Width="300">
				<ItemsControl Theme="{StaticResource ItemsControlTheme}">
					<TextBlock Text="4444" />
					<TextBlock Text="5555" />
					<TextBlock Text="6666" />
				</ItemsControl>
			</Border>
		</StackPanel>
	</Grid>
</Border>
```
</details>
<details>
  <summary>3. MainView code-behind:</summary>
  
```cs
private TranslateTransform _translate;

public MainView()
{
    InitializeComponent();

    Panel.RenderTransform = _translate = new TranslateTransform
    {
        Transitions = new Avalonia.Animation.Transitions
        {
            new DoubleTransition
            {
                Duration = System.TimeSpan.FromMilliseconds(300),
                Property = TranslateTransform.XProperty
            }
        }
    };
}

private void Left_Click(object? sender, Avalonia.Interactivity.RoutedEventArgs e)
{
    _translate.X = 0;
}

private void Right_Click(object? sender, Avalonia.Interactivity.RoutedEventArgs e)
{
    _translate.X = -310;
}
```
</details>

<details>
  <summary>4. ItemsControlTheme:</summary>
  
```xml
<ControlTheme x:Key="ItemsControlTheme"
			  TargetType="ItemsControl">
	<Setter Property="ScrollViewer.HorizontalScrollBarVisibility" Value="Disabled" />
	<Setter Property="ScrollViewer.VerticalScrollBarVisibility" Value="Hidden" />
	<Setter Property="ScrollViewer.IsScrollChainingEnabled" Value="True" />
	<Setter Property="ScrollViewer.IsScrollInertiaEnabled" Value="True" />
	<Setter Property="ItemsPanel">
		<ItemsPanelTemplate>
			<VirtualizingStackPanel />
		</ItemsPanelTemplate>
	</Setter>
	<Setter Property="Template">
		<ControlTemplate>
			<Border Name="border"
					ClipToBounds="{TemplateBinding ClipToBounds}"
					Background="{TemplateBinding Background}"
					BorderBrush="{TemplateBinding BorderBrush}"
					BorderThickness="{TemplateBinding BorderThickness}"
					CornerRadius="{TemplateBinding CornerRadius}">
				<ScrollViewer Name="PART_ScrollViewer"
							  VerticalSnapPointsType="{TemplateBinding (ScrollViewer.VerticalSnapPointsType)}"
							  HorizontalSnapPointsType="{TemplateBinding (ScrollViewer.HorizontalSnapPointsType)}"
							  HorizontalScrollBarVisibility="{TemplateBinding (ScrollViewer.HorizontalScrollBarVisibility)}"
							  VerticalScrollBarVisibility="{TemplateBinding (ScrollViewer.VerticalScrollBarVisibility)}"
							  IsScrollChainingEnabled="{TemplateBinding (ScrollViewer.IsScrollChainingEnabled)}"
							  IsScrollInertiaEnabled="{TemplateBinding (ScrollViewer.IsScrollInertiaEnabled)}"
							  AllowAutoHide="{TemplateBinding (ScrollViewer.AllowAutoHide)}"
							  BringIntoViewOnFocusChange="{TemplateBinding (ScrollViewer.BringIntoViewOnFocusChange)}">
					<ItemsPresenter Name="PART_ItemsPresenter"
									ItemsPanel="{TemplateBinding ItemsPanel}"
									Margin="{TemplateBinding Padding}"/>
				</ScrollViewer>
			</Border>
		</ControlTemplate>
	</Setter>
</ControlTheme>
```
</details>

**Expected behavior**
All items from ItemsControls are rendered

**Screenshots**

https://github.com/AvaloniaUI/Avalonia/assets/23280622/adf44924-f775-4103-a943-c53bc3f460b5

**Desktop (please complete the following information):**

- OS: Windows
- Version: 11.0.2

**Additional context**
VirtualizingStackPanel render other items when visual is invalidated. For example, button change state from `pointerover` to `normal` or window resized etc. Also all right if use StackPanel instead VirtualizingStackPanel.

**Comment by timunie** [comment]
@Meloman19 I think LayouTransform is what you need. Render transform will not have enough information about transformed bounds. 

https://docs.avaloniaui.net/docs/next/reference/controls/layouttransformcontrol#more-information

**Comment by Meloman19** [comment]
@timunie Yes, LayoutTransform is works, but losing the benefit of RenderTransform.

As I can see from the source code, RenderTransform used on bounds transform. But main problem in EffectiveViewport. When panel is out of bounds, value of calculated viewport becomes empty and VirtualizedStackPanel prepare only one entity. And after change render position not raised EffectiveViewport event, because it's part of layout pass.

Instead of using an EffectiveViewport event, VirtualizingPanel can use directly parent IScrollable interface. But in this case invalidating event is missing. Or move EffectiveViewport to render pass...

**Comment by timunie** [comment]
I guess this is by design, but will ask @grokys first

**Comment by abirne** [comment]
@grokys, As RenderTransform are independent of the layout system by definition, we must not interconnect the layout-level event (`EffectiveViewportChanged`) to it, Otherwise we must consider to invalidate the effective viewport calculation on its changes as one of key parameters... I think this issue is wider than virtualizing items controls and this event generally is faulty, now. Therefore hinting the items control is not a general solution to resolve this issue.

**Comment by abirne** [comment]
@grokys, I mean: it seems the viewport caculation should be considered independent from RenderTransform, Otherwise, Its Changes should cause viewport invalidation, because it is one of the main parameters that affects it.

**Comment by timunie** [comment]
@abirne if you have any ideas, you can also clone source and try them. You can also open PRs if you want to and have a possible solution.

**Comment by abirne** [comment]
> [@abirne](https://github.com/abirne) if you have any ideas, you can also clone source and try them. You can also open PRs if you want to and have a possible solution.

Ok. But accroding to the [contributing guidelines](https://github.com/AvaloniaUI/Avalonia/blob/master/CONTRIBUTING.md),  I must share my opinion and disscuss with core team before creating a PR (in this case), to ensure the idea is acceptable and actually is currect and matches plans for making breaking changes... some issues resolving need deeper think and collaboration.

**Comment by timunie** [comment]
You can also make a draft PR for testing and discussion. There is no guarantee it will be merged but can really help both parties. But don't feel forced.

**Comment by grokys** [comment]
The current design for `EffectiveViewportChanged` is taken from UWP/WInUI and iirc, the [`EffectiveViewport`](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.frameworkelement.effectiveviewportchanged?view=winrt-26100#effectiveviewport) there also includes transforms, so I'm wondering whether WinUI also has similar problems.

I think the first thing to check is:

1. That we do have the same semantics/behavior as WinUI
2. Whether WinUI also has the same issue

If we have different behavior to WinUI, then we should fix our implementation.

If have the same behavior as WinUI, and WinUI also has the same problems, then we may want to consider adding an `UntransformedEffectiveViewport` or `EffectiveLayoutViewport` (naming TBD) property to the event args.

**Comment by abirne** [comment]
> ... there also includes transforms, so I'm wondering whether WinUI also has similar problems.
...

@grokys, Looking at the WinUI source code, yes; it seems transforms are taken into account, and logically the same problem must exist in WinUI (although I didn't have enough time to create a test project to reproduce and ensure there is not any layout invalidation on render transform change, which seems to be not done according to the official documentation).

Finally, `LayoutManager.UpdateLayout` causes the `EffectiveViewportChanged` event to be raised if needed and checks a transform-dirty flag on elements in the process. But it seems the invalidation and calculation process is a bit different in WinUI than Avalonia...

> ... then we may want to consider adding an UntransformedEffectiveViewport or EffectiveLayoutViewport (naming TBD) property to the event args.

This change itself will not resolve the problem unless the effective viewport event firing is intended to be done after any change of the render transform when there is not any other change in layout-related things...

**Comment by abirne** [comment]
In short, if it is intended to apply the render transform effect on EffectiveViewport, then I can suggest:

- using a kept state flag in each visual that must be invalidated immediately on change of related things to the Effective Viewport calculation (such as `RenderTransform` property).

- later, calculating a part of the effective viewport for the hierarchy of related visuals upward (if needed), probably cached ~~in the render thread~~ (partially usable in next calculations). and raising the related event in a proper time on the UI thread.

-  reflecting the effective viewport in two separate forms, `EffectiveLayoutViewport` and `EffectiveViewport` (probably through EffectiveViewportChangedEventArgs) to be usable for different purposes such as virtualization and hit-testing, etc.
